### ISSUE TITLE
Frames cannot display continuously using MediaController 

### ISSUE THREAD

**Issue by speedpursuer** [koral--/android-gif-drawable#301]
When continuously dragging the progress bar, imageView becomes black with nothing showing and frame comes to appear only after the the drag stops. I expect the play is smooth when manipulating the progress bar.

I also tried to create an UI interaction myself by calling seekToFrameAndGet(). It works without the screen flash problem (black screen) but the play is still not smooth.

I'm also wondering if GifTextureView is a performant solution for this use case. But I didn't find a way for GifTextureView to work with Drawable to manipulate frames. Would you please help advise?

Thank you!

**Comment by koral--** [comment]
`GifTextureView` can only play GIFs, it does not support seeking.

I would implement smooth seeking with `GifTexImage2D` here is an example:
https://github.com/koral--/android-gif-drawable-sample/blob/4ba7a8a21cf2dfb79a843296610c4d0a89ef5dae/sample/src/main/java/pl/droidsonroids/gif/sample/GifTexImage2DFragment.java

**Comment by speedpursuer** [comment]
Thanks Karol for your quick response. Yes, the demo looks great. Very smooth for the forward. Just a little problem with the rewind. The screen is flashing during rewind. If this can be refined, the seeking is perfect to me.

Thanks again!

**Comment by speedpursuer** [comment]
FYI - the screenshot of rewind problem I can see.
[http://i1.hoopchina.com.cn/blogfile/201605/18/BbsImg146353740639301_143x251.gif](url)

**Comment by koral--** [comment]
It seems that it is issue similar to #267.
You can synchronize rendering with seeking like here: https://github.com/koral--/android-gif-drawable-sample/commit/b7a33607c34aa6e04421536bfe957c22913ece25
or (better) implement some debouncing. In other words discard some seek requests if another one is in progress, just like [throttleLast from RxJava](http://reactivex.io/RxJava/javadoc/rx/Observable.html#throttleLast(long,%20java.util.concurrent.TimeUnit)) (or similar operator) works.

**Comment by speedpursuer** [comment]
Thanks for your swift replay. Your synchronize code work perfect! 

Thank you very much for the help. Android-gif-drawable is great btw ; )

**Comment by speedpursuer** [comment]
Just a suggestion - would it be good if play() and stop() are encapsulated in GifTexImage2D? (although those can be accomplished in client code using seekToFrame() and getFrameDuration() )

**Comment by speedpursuer** [comment]
I'm sorry I have one more question - the aspect ratio of frame rendered by openGL is changed ( both the width and height fit the screen). If I want to keep the aspect ratio as it is and have just one dimension (the longer one) fit, would you help suggest how this can be implemented? 
Thank you!

**Comment by koral--** [comment]
`play()` functionality currently exists as `startDecoderThread()` and `stop()` as `stopDecoderThread()`. To keep aspect ratio you can set size (layout width and height) of your `GLSurfaceView` appropriately or  use OpenGL APIs to transform texture.

**Comment by speedpursuer** [comment]
Thank you! Yes, setting layout of GLSurfaceView works. For the startDecoderThread(), I found the some noisy white color lines here and there on the screen while playing. It's very similar to the symptom where manually seeking is performed without synchronize.

**Comment by koral--** [comment]
Reason of that is the same - seeking is not synchronized with drawing. Currently `startDecoderThread()` spawns a native thread so there is no way to add synchronization without modifying library sources.

However there is a plan to move all threads from native to java code, so library users like you will be able to extend library classes easily (eg. add synchronization wherever needed).

**Comment by speedpursuer** [comment]
That sounds great! Look forward to that feature.

I just found another issue with glTexImage2D - the gif image can not display in some relatively larger screen devices. Using the same sample code, it works on a device of screen 900 x 600 but only shows black screen for GifTexImage2DFragment on a 1200 x 800 device which supports GLES2. 

I found some GLES OOM errors from console:

<program> could not be made part of current state. <program> is not linked (gles2_program_get_attrib_location at hardware/rk29/maliT760/driver/product/gles/src/program/mali_gles2_program_api.c:1133)

<index> is greater than or equal to GL_MAX_VERTEX_ATTRIBS (gles_vertexp_attrib_pointer_common at hardware/rk29/maliT760/driver/product/gles/src/vertex/mali_gles_vertex_internal.c:171)

there is no current program object (gles2_programp_set_uniform at hardware/rk29/maliT760/driver/product/gles/src/program/mali_gles2_program_internal.c:885)

Can you please help give some hints what's going wrong? Thank you very much!

**Comment by koral--** [comment]
Does adding precision to fragment shader varying variable fix the problem? Namely try changing this line: https://github.com/koral--/android-gif-drawable-sample/blob/master/sample/src/main/java/pl/droidsonroids/gif/sample/GifTexImage2DFragment.java#L64 to the following form: `varying mediump vec2 textureCoordinate;`

If not please attach shader information log. It may be obtained by adding line like that: `Log.w("shader info log", glGetShaderInfoLog(shader));` after `glCompileShader()` call, here: https://github.com/koral--/android-gif-drawable-sample/blob/master/sample/src/main/java/pl/droidsonroids/gif/sample/GifTexImage2DFragment.java#L146
Log should appear in logcat.

**Comment by speedpursuer** [comment]
Yes, that does help and the image can show up. Thanks!

And I also noticed for some middle size gifs (e.g. https://media.giphy.com/media/3oEjHVoG0nr2O3SC4M/giphy.gif), the seeking is not quite smooth with GifTexImage2D. 

I don't think it's because of the image size - if you try this (https://media.giphy.com/media/3oEjHW5syYCSIR7f6U/giphy.gif), the seeking has no problem.

Do you think if there is some tuning opportunity to have better performance for the seeking?

Have a nice weekend!

**Comment by koral--** [comment]
I'll add `mediump` to the sample code.
I'll check what can be done about seeking performance.

**Comment by speedpursuer** [comment]
Yeah, I didn't see any drawback using mediump. Look forward to the improvement!

**Comment by speedpursuer** [comment]
BTW, for your suggestion for keeping the image aspect ratio, would you please include some example code using OpenGL APIs to transform texture in the sample project if possible? I think that will be helpful for the lib users like me to use GifTexImage2D out of box.

Thanks a lot for your support!

**Comment by koral--** [comment]
OK, I'll try to add it ASAP.

**Comment by koral--** [comment]
I've modified GifTexImage2D example in sample project. Now GIF is centered in surface and keeps its aspect ratio: https://github.com/koral--/android-gif-drawable-sample/commit/81c926167f5fedcc597d862aa95ca9b55b9580b4

**Comment by speedpursuer** [comment]
Thanks for update! I tested it on two devices and there seems to be some display issue - the gif is centered but some other things are also showing around the gif while it is playing. I recorded the screen and made it as a gif fyi - http://i.giphy.com/xT8qBtqRX91L4NbvkQ.gif

Another thing I would suggest is that the image scale type is better to be CENTER_CROP (Scales the image so that both dimensions will be greater than or equal to the corresponding dimension of the screen. At least one dimension (width or height) will fit exactly. Image is centered within screen's bounds.) which would look like more nature user experience.

Below is my code used to do CENTER_CROP scale with GLSurfaceView FYI:

private void setLayout(GLSurfaceView view) {
        float imageWidthToHeight = ((float) mGifTexImage2D.getWidth()) / ((float) mGifTexImage2D.getHeight());
        float viewWidthToHeight = (float)screenWidth / (float)screenHeight;
        float top = 0, left = 0, imageViewWidth, imageViewHeight;

        //If viewWidthToHeight > imageWidthToHeight, the image is taller than the screen, so the image height = screen height. left is left and right margin to the screen's bounds.
        if(viewWidthToHeight > imageWidthToHeight) {
            imageViewWidth = screenHeight * imageWidthToHeight;
            imageViewHeight = screenHeight;
            left = (screenWidth - imageViewWidth) / 2;
        }else {
            imageViewWidth = screenWidth;
            imageViewHeight = screenWidth / imageWidthToHeight;
            top = (screenHeight - imageViewHeight) / 2;
        }

        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams((int)imageViewWidth, (int)imageViewHeight);
        layoutParams.setMargins((int)left, (int)top, (int)left, (int)top);
        view.setLayoutParams(layoutParams);
    }

**Comment by koral--** [comment]
OK, those other things showing around proper image are duplicated pixels from border. They appear due to GL_TEXTURE_WRAP_S/T parameters.
It seems that center crop mode can be achieved using scaling and translation only. Wrapping won't occur because entire surface will be covered by GIF.
I'll update the sample.

**Comment by speedpursuer** [comment]
Thanks and good to know the cause is found. 

Yes, I hope center crop mode is achieved through openGL/GifTexImage2D as well. Changing the surface view size (the method I showed above) introduces some limitations (e.g. for user interaction with the view).

Look forward to you update!

**Comment by koral--** [comment]
Sorry for the delay.
Tearing (issue from https://github.com/koral--/android-gif-drawable/issues/301#issuecomment-219903866) should not occur in `1.1.17-SNAPSHOT` synchronization has been added.
I keep in mind other questions.

**Comment by speedpursuer** [comment]
Thank you! I've tested the startDecoderThread() and stopDecoderThread(). They work with no tearing issue. 

Look forward to further update for the seeking performance and scaling with opengl

**Comment by koral--** [comment]
Sorry again for the delay.
I've done some investigations about seeking and found no better way than introduced in https://github.com/koral--/android-gif-drawable/blob/42036e2b4fee6b03a708a7e3e03a99dd1d87f68b/src/main/jni/control.c
There may be no key-frames in GIF and in some cases all frames from current or even the beginning to desired one need to be rendered. This is exactly the case in your first GIF - entire frames has transparency flags set and no disposal method (except last one - which has no meaning in seeking). On the other hand the second GIF has all the frames covering entire canvas and disposable to background color (except last one), so there is no need to render any additional frames when seeking.

See here: https://github.com/koral--/android-gif-drawable/issues/274#issuecomment-210098076 and issue linked there for detailed explanation.

I keep in mind center crop scaling.

**Comment by koral--** [comment]
One of the ways to achieve center crop scaling is to use translation and scaling of model matrix like in this example: https://github.com/koral--/android-gif-drawable-sample/blob/0a4bb29db3bea7e0af06160f3feb038770a0e98c/sample/src/main/java/pl/droidsonroids/gif/sample/Image.java#L129

**Comment by speedpursuer** [comment]
Thanks a lot for the updates! 

I can understand the reason for the seeking issue. I will also try the center crop scaling by referring your example and let your know the result.

Thanks again for all your kind help!
